crates/store/re_video/src/decode/av1.rs

```diff
@@ -0,0 +1,408 @@
+        self.comms
+            .num_outstanding_resets
+            .fetch_add(1, Ordering::SeqCst);
+        self.command_tx.send(Command::Reset).ok();
+    }
+
+    /// Blocks until all pending frames have been decoded.
+    // NOTE: The interface is all `&mut self` to avoid certain types of races.
+    pub fn flush(&mut self) {
+        re_tracing::profile_function!();
+        let (tx, rx) = crossbeam::channel::bounded(0);
+        self.command_tx.send(Command::Flush { on_done: tx }).ok();
+        rx.recv().ok();
+    }
+}
+
+impl Drop for Decoder {
+    fn drop(&mut self) {
+        re_tracing::profile_function!();
+        self.comms.should_stop.store(true, Ordering::SeqCst);
+        self.command_tx.send(Command::Stop).ok();
+        // NOTE: we don't block here. The decoder thread will finish soon enough.
+    }
+}
+
+type OutputCallback = dyn Fn(Result<Frame>) + Send + Sync;
+
+fn create_decoder() -> Result<dav1d::Decoder, dav1d::Error> {
+    re_tracing::profile_function!();
+
+    let mut settings = dav1d::Settings::new();
+    settings.set_strict_std_compliance(false);
+    settings.set_max_frame_delay(1);
+
+    dav1d::Decoder::with_settings(&settings)
+}
+
+fn decoder_thread(comms: &Comms, command_rx: &Receiver<Command>, on_output: &OutputCallback) {
+    #![allow(clippy::debug_assert_with_mut_call)]
+
+    let mut decoder = match create_decoder() {
+        Err(err) => {
+            on_output(Err(Error::Initialization(err)));
+            return;
+        }
+        Ok(decoder) => decoder,
+    };
+
+    while let Ok(command) = command_rx.recv() {
+        if comms.should_stop.load(Ordering::SeqCst) {
+            re_log::debug!("Should stop");
+            return;
+        }
+
+        // If we're waiting for a reset we should ignore all other commands until we receive it.
+        let has_outstanding_reset = 0 < comms.num_outstanding_resets.load(Ordering::SeqCst);
+
+        match command {
+            Command::Chunk(chunk) => {
+                if !has_outstanding_reset {
+                    submit_chunk(&mut decoder, chunk, on_output);
+                    // NOTE: in all video's I've tested, there is one frame per chunk
+                    let _num_frames = output_frames(&comms.should_stop, &mut decoder, on_output);
+                }
+            }
+            Command::Flush { on_done } => {
+                debug_assert!(matches!(decoder.get_picture(), Err(dav1d::Error::Again)),
+                    "There should be no pending pictures, since we output them directly after submitting a chunk.");
+
+                on_done.send(()).ok();
+            }
+            Command::Reset => {
+                decoder.flush();
+
+                debug_assert!(matches!(decoder.get_picture(), Err(dav1d::Error::Again)),
+                    "There should be no pending pictures, since we output them directly after submitting a chunk.");
+
+                comms.num_outstanding_resets.fetch_sub(1, Ordering::SeqCst);
```

Comment:
```suggestion
                comms.num_outstanding_resets.fetch_sub(1, Ordering::Release);
```
.. but since we're only syncing with ourselves this might actually be `Relaxed`, there's no state to publish here that we need. This one time I'd also go with better safe than sorry ;-)